odoo/models.py

```diff
@@ -5550,8 +5550,13 @@ def __len__(self):
 
     def __iter__(self):
         """ Return an iterator over ``self``. """
-        for id in self._ids:
-            yield self._browse(self.env, (id,), self._prefetch_ids)
+        if len(self._ids) > PREFETCH_MAX and self._prefetch_ids is self._ids:
```

Comment:
just a small question: why does this apply only if the prefetch set is the same as the recordset? is there an actual technical reason or just an use case not worth supporting?

Reply:
It's all about having **predictable performance**.

Consider the following code:
```python
records = model.search(...)
for record in records:              # optimization applies here
    print(record.name)
    for line in record.line_ids:    # optimization does not apply here
        print(line.name)
```
The optimization applies to all recordsets returned by `browse()`.  This includes the result of `search()`, `create()`, `self` in compute methods, `self` in RPC-called methods, etc.  So, in the example above, `record._prefetch_ids` is guaranteed to be no more than `PREFETCH_MAX` ids, which reduces the overhead of the prefetching of `record.name` to the minimum.

However, the situation is quite different for *secondary* recordsets, like `record.line_ids`.  The value of `record.line_ids._prefetch_ids` is an *iterable* object that represents the union of the values of `line_ids` on `record._prefetch_ids` in cache, and the iteration on that object is much more costly!  Given that the arity of the relation `line_ids` is unknown, the cost of iterating on `record.line_ids._prefetch_ids` is essentially not predictable, and slicing its result is tricky, as the size of `record.line_ids` may go from zero to several batches.  Moreover, at the time of `__iter__`, the ORM doesn't even know whether `line._prefetch_ids` will actually be used.

So, for `line.name`, the ORM will iterate on `line._prefetch_ids = record.line_ids._prefetch_ids` to determine prefetching candidates.  That case has potentially more overhead, but the top-level optimization (iteration on `records`) keeps it under relative control.  Indeed, the size of `record.line_ids._prefetch_ids` is around `average_arity_of_line_ids * PREFETCH_MAX`.

Long story short, some nested iterations *may* require some manual optimization, but those cases are much less common than simple loops.

Reply:
Thanks for the explanation, I guess it wasn't such a small question after all :rofl: at least it's documented in the PR now!